crates/solver/src/settlement.rs

```diff
@@ -299,6 +318,77 @@ impl Settlement {
+        }
+        // The following check is quadratic in run-time, although a similar check with linear run-time would also be possible.
+        // For the linear implementation, one would have to find a unique scaling factor that scales the external prices into
+        // the settlement prices. Though, this scaling factor is not easy to define, if reference tokens like ETH are missing.
+        // Since the checks would heavily depend on this scaling factor, and its
+        // derivation is non-trivial, we decided to go for the implementation with quadratic run time. Settlements
+        // will not have enough tokens, such that the run-time is important.
+        self
+            .clearing_prices()
+            .iter()
+            .combinations(2)
+            .all(|clearing_price_vector_combination| {
+                let (sell_token, sell_price) = clearing_price_vector_combination[0];
+                let clearing_price_sell_token = sell_price.to_big_rational();
+                let (buy_token, buy_price) = clearing_price_vector_combination[1];
+                let clearing_price_buy_token = buy_price.to_big_rational();
+
+                if matches!(tokens_to_satisfy_price_test, TokensForPriceCheck::Tokens(token_list) if (!token_list.contains(sell_token)) || !token_list.contains(buy_token))
+                {
+                    return true;
+                }
+                let external_price_sell_token = match external_prices.price(sell_token) {
+                    Some(price) => price,
+                    None => return true,
+                };
+                let external_price_buy_token = match external_prices.price(buy_token) {
+                    Some(price) => price,
+                    None => return true,
+                };
+                // Condition to check: Deviation of clearing prices is bigger than max_settlement_price deviation
+                //
+                // |clearing_price_sell_token / clearing_price_buy_token - external_price_sell_token/external_price_buy_token)|
+                // |----------------------------------------------------------------------------------------------------------|
+                // |                     clearing_price_sell_token / clearing_price_buy_token                                  |
+                // is bigger than:
+                // max_settlement_price_deviation
+                //
+                // This is equal to: |clearing_price_sell_token*external_price_buy_token-external_price_sell_token*clearing_price_buy_token|>
+                // max_settlement_price_deviation*clearing_price_buy_token*external_price_buy_token *clearing_price_sell_token
+
+                let price_check_result = clearing_price_sell_token
+                    .clone()
+                    .mul(external_price_buy_token)
+                    .sub(&external_price_sell_token.mul(&clearing_price_buy_token)).abs()
+                    .lt(&max_settlement_price_deviation
```

Comment:
Sneaking an inequality into a chain of other arithmetic operations is misleading for me, I was not expecting the result to be Boolean on the next line

I think this would be more clear if the inequality was made in the if statement below.

Reply:
I kind of like the succinctness of it

Reply:
I hadn't realized before, but the `add`, `mul`, etc. operations are using the `std::ops::*` traits which means they are what get called with `+`, `*`, etc. (so no need to use the method calls). This means you could have written this as:

```rust
    let price_check_result =
        (clearing_price_sell_token * &external_price_buy_token - &external_price_sell_token * &clearing_price_buy_token).abs()
        < max_settlement_price_deviation;
```

Note that in this case, I also see the value of writing it in terms of the `add`, `mul`, etc. method calls and am  happy either way.